### DISCUSSION TITLE
next/font/google download timeout should be configurable

### DISCUSSION THREAD

**Discussion by carloscarnero** [vercel/next.js#47009] [Ideas]
### Goals

1. In development mode, if I'm not mistaken, downloading a font from Google [times out after 3 seconds](https://github.com/hanneslund/next.js/blob/5b1ce236d56e94fde500e40654ba9885df08e371/packages/font/src/google/fetch-css-from-google-fonts.ts#L33). This value should be configurable to account for a variety of environments.

### Background

In Next.js release v13.2.4 a timeout for the process of downloading fonts from Google Fonts was merged in #46834 for development environments. While the idea seems solid, when the network is flaky, congested, or otherwise crappy, this timeout fires too soon and fonts can't be downloaded (in fact, this has happened to me twice already.)

### Proposal

The value for the timeout should be configurable, while keeping the chosen 3000 milliseconds as default. Honestly, I have no idea if the override should be set via environment variables or in `next.config.js`.

**Comment by carloscarnero** [comment]
@hanneslund 👋 do you think this have any merit?

**Comment by ReoHakase** [comment]
Hi  dear @hanneslund 

The current `3000` timeout setting is not sufficient for CJK fonts most of the time. (At least for my average network environment)
https://github.com/vercel/next.js/issues/45080#issuecomment-1535150448

The suggested configurable timeout would be great for DX, especially for those who develop apps in CJK languages.
Could you consider to implement the mentioned feature?

Thanks in advance

**Comment by darioscattolini** [comment]
I'm having the same issue with a regular Latin font with just two font weights, and no particular network issue: the request times out and fails. I'm not sure if making it configurable is the best solution, but at least a longer timeout could be used.

Perhaps PR #51890 solves the issue.

**Comment by absolutegravitas** [comment]
Same issue and production build fails because of it.... is this on the roadmap for "fix" anytime soon otherwise use of next/font is useless for our projects

**Comment by SC0d3r** [comment]
I am facing this issue when downloading google fonts, there should be an option to increase/set the timeout when downloading google fonts in next/font

**Comment by absolutegravitas** [comment]
Going to go back to manual script references... this package is garbage

**Comment by sneko** [comment]
@styfle  @timneutkens @berlysia  following your contributions to the font package (https://github.com/vercel/next.js/pull/58075 and others to retry fetching), any chance to have a configurable timeout? In my case the provider build environment I have always fails to fetch the font within the default timeout (it's a known issue, with people behind VPN...).

Thank you,

**Comment by allenchuang** [comment]
Facing this similar issue. following

**Comment by allenchuang** [comment]
![image](https://github.com/vercel/next.js/assets/4248780/60e73a45-4335-41a3-aa9d-42ad156ed6d2)
bump, what is the fix here? I find myself often need to redepoly multiple times in order to get a green build.

**Reply by crobley** [reply]
> ![image](https://private-user-images.githubusercontent.com/4248780/339151810-60e73a45-4335-41a3-aa9d-42ad156ed6d2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0NDgyMTMsIm5iZiI6MTcyNzQ0NzkxMywicGF0aCI6Ii80MjQ4NzgwLzMzOTE1MTgxMC02MGU3M2E0NS00MzM1LTQxYTMtYWE5ZC00MmFkMTU2ZWQ2ZDIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjdUMTQzODMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWFkMDE5MDIxYjQzMjVlNmVlNTY2ZTgzMDFlYzM0ODkxZGY0MjRlNjUwNDFhMjk4MWJkN2E3YTNhOTFlNDRiZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.k19sNDpXjrYBMcPMbT8XNyta4dhNliHVTsYsez0yj-c) bump, what is the fix here? I find myself often need to redepoly multiple times in order to get a green build.

We have the same issue running on a GitLab runner. Sometimes it's the first time, others upwards of 20 times!

**Comment by Harris-Ferguson** [comment]
Facing this issue as well, we have CI/CD builds fail pretty consistently due to this. Following.

**Comment by Luise8** [comment]
So this is still happening with github actions. I have to run many times the jobs. I'm not behind a VPN...

**Comment by zlatiz** [comment]
Running into this issue as well while using Digital Ocean App services. I have to run the CI/CD at least 10 - 15 times...

**Comment by maggie44** [comment]
It seems common in GitHub Actions when building arm64 images which require emulation. It's consumes a lot of resources so takes longer to perform actions like this, which then times out regularly.

**Reply by anthonyalayo** [reply]
Exactly my situation.

**Comment by queenvictoria** [comment]
Our use case is for a multilingual site where we consume 20+ Google fonts. The 3000ms timeout means that 9/10 development runs _and_ 9/10 production builds fails rendering `next/fonts/google` unusable from a dx point of vieww. A configurable timeout would go a large distance to resolving this for us.

**Comment by mhornbacher** [comment]
Just to pile on I'm observing a fun behavior in docker where after 3 seconds it fails for all 3 retries instantly. This also occurs 100% of the time if you use a large font such as Simplified Chinese (which our app requires and is a 17MB TTF file)

**Comment by whoisjuan** [comment]
Just here to say I also have this problem. Has anyone figured out a solution?

**Reply by mhornbacher** [reply]
We settled on using on primary font (our latin font) which downloads generally just fine. While having the heavier fonts such as Mandarin get pulled on the web-page the old fashioned way. This causes a minor one-time shift for non-latin languages but resolved the issue for us :). Its a very anglo-centric solution for sure but it works for our needs.

**Comment by hashilfinstreet** [comment]
I am facing the same issue

  ▲ Next.js 14.2.7
  - Environments: .env

   Creating an optimized production build ...
request to https://fonts.googleapis.com/css2?family=Nunito:wght@200..1000&display=swap failed, reason: 

Retrying 1/3...
request to https://fonts.googleapis.com/css2?family=Nunito:wght@200..1000&display=swap failed, reason: 

Retrying 2/3...
request to https://fonts.googleapis.com/css2?family=Nunito:wght@200..1000&display=swap failed, reason: 

Retrying 3/3...
FetchError: request to https://fonts.googleapis.com/css2?family=Nunito:wght@200..1000&display=swap failed, reason: 
    at ClientRequest.<anonymous> (/node_modules/next/dist/compiled/node-fetch/index.js:1:66160)
    at ClientRequest.emit (node:events:524:28)
    at emitErrorEvent (node:_http_client:101:11)
    at TLSSocket.socketErrorListener (node:_http_client:504:5)
    at TLSSocket.emit (node:events:524:28)
    at emitErrorNT (node:internal/streams/destroy:169:8)
    at emitErrorCloseNT (node:internal/streams/destroy:128:3)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
    at runNextTicks (node:internal/process/task_queues:64:3)
    at process.processTimers (node:internal/timers:516:9) ***
  type: 'system',
  errno: 'ETIMEDOUT',
  code: 'ETIMEDOUT'
***
Failed to compile.

src/app/fonts.ts
`next/font` error:
Failed to fetch `Nunito` from Google Fonts.


> Build failed because of webpack errors
The command '/bin/sh -c npm i && npm run build' returned a non-zero code: 1

Error: Process completed with exit code 1.
